### PR TITLE
Implement vdisk rollback

### PR DESCRIPTION
**Pull request by khaledkbadr**
related to #430

Signed-off-by: Khaled K. Badr <khaledkbadr@outlook.com>

### CODE REVIEW

File: templates/storage_cluster/actions.py [hunk 1]
```diff
@@ -30,6 +30,16 @@ def get_cluster(service):
...
 
 
+def get_baseport(node):
+    ports = node.client.info.port()
+    baseport = 2000
+    for portInfo in ports:
+        port = portInfo.get('port', 0)
+        if str(port).startswith('200') and port >= baseport:
+            baseport = port + 1
+    return baseport
```
Comment (grimpy) [inline templates/storage_cluster/actions.py]: please redo this withouth string.
for instnace make a list of used tcp ports and do a while loop to find first avaialble nr above 2000

File: api/vdisk/VdiskRollback.go [hunk 2]
```diff
@@ -5,7 +5,7 @@ import (
 )
 
 type VdiskRollback struct {
-	Epoch int `yaml:"epoch" json:"epoch" validate:"nonzero"`
+	Epoch uint64 `yaml:"epoch" json:"epoch" validate:"nonzero"`
```
Comment (grimpy) [inline api/vdisk/VdiskRollback.go]: deal with the nanoseconds in the actor not in the api.

File: templates/vdisk/actions.py [hunk 3]
```diff
@@ -144,10 +151,51 @@ def pause(job):
 
 
 def rollback(job):
+    import random
+    import yaml
     service = job.service
     service.model.data.status = 'rollingback'
-    # TODO: rollback disk
-    service.model.data.status = 'running'
+    ts = job.model.args['timestamp']
```
Comment (grimpy) [inline templates/vdisk/actions.py]: there should be a check here on the disk type. and disk status

File: templates/zerodisk/actions.py [hunk 4]
```diff
@@ -163,7 +166,8 @@ def start_tlog(service, container, config):
     container.client.filesystem.upload(configpath, configstream)
     if not is_job_running(container, cmd='/bin/tlogserver'):
         port = get_tlog_port(container)
-        container.client.system('/bin/tlogserver -address 0.0.0.0:{} -config {} -k {} -m {}'.format(port, configpath, k, m))
+        logpath = '/tlog_{}.log'.format(service.name)
+        container.client.system('/bin/tlogserver -address 0.0.0.0:{} -config {} -k {} -m {} -logfile {}'.format(port, configpath, k, m, logpath))
```
Comment (grimpy) [inline templates/zerodisk/actions.py]: tlogserver shouldnt listen on 0.0.0.0 but instead onf node.storageaddr

File: templates/vdisk/actions.py (hunk 3 above)
Comment (khaledkbadr) [inline templates/vdisk/actions.py]: The check is in the process change method

File: templates/nbdserver/actions.py [hunk 5]
```diff
@@ -134,42 +133,12 @@ def install(job):
     else:
         # send a siganl sigub(1) to reload the config in case it was changed.
         job = is_job_running(container)
+        print(job)
```
Comment (grimpy) [inline templates/nbdserver/actions.py]: why is this print here?

File: templates/nbdserver/actions.py [hunk 6]
```diff
@@ -193,7 +162,7 @@ def stop(job):
 
     # Delete tmp vdisks
     for vdiskservice in vdisks:
-        vdiskservice.model.data.status = 'halted'
+        j.tools.async.wrappers.sync(vdiskservice.executeAction('pause'))
         vdiskservice.saveAll()
```
Comment (grimpy) [inline templates/nbdserver/actions.py]: this save shoudlnt be required anymore please remove and test again

File: templates/nbdserver/actions.py [hunk 7]
```diff
@@ -223,7 +192,7 @@ def monitor(job):
     running = is_job_running(get_container(service, get_jwt_token(job.service.aysrepo)))
     for vdisk in vdisks:
         if running:
-            vdisk.model.data.status = 'running'
+            j.tools.async.wrappers.sync(vdisk.executeAction('start'))
         else:
-            vdisk.model.data.status = 'halted'
+            j.tools.async.wrappers.sync(vdisk.executeAction('pause'))
     vdisk.saveAll()
```
Comment (grimpy) [inline templates/nbdserver/actions.py]: this save is probably not required anymore

File: templates/tlogserver/actions.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                -logfile {log} \
+                -config {config}'
+                .format(ip=ip,
+                        port=port,
+                        config=configpath,
+                        k=k,
+                        m=m,
+                        log=logpath)
+            )
+        if not is_job_running(container, cmd='/bin/tlogserver'):
```
Comment (grimpy) [inline templates/tlogserver/actions.py]: check on on listening port not on the binary itself to ensure its fully started

**Pull request merged**